### DISCUSSION TITLE
Customize loading spinner position and appereance

### DISCUSSION THREAD

**Discussion by danielvodnanskyextconradcom** [apache/superset#28768] [Ideas]
## Screenshot

![image](https://github.com/apache/superset/assets/171031071/75d2bf44-b966-429d-b1bc-fdc640fee6fa)

## Description

This spinner (loader) cannot be customized. It's always positioned in the middle of iframe, so user on our page sometimes don't see something is happenning if the iframe is in lower part and they see just part of is (like a grey rectangle.)

We cannot apply styles on it, because they are loaded with the content itself after the spinner dissapear.

## Design input
I suggest the possibility to reposition the spinner (set the top/left percent position) OR display multiple spinners OR add custom style to the spinner OR modify directly its html. Also it would be great to be able to upload custom spinner.

**Comment by rusackas** [comment]
All of the things you're asking for here could be contributed in a PR as feature additions (positioning improvements, additional classes needed for CSS, or even a customizable path in the config for a custom image, as we do in with the app logo). I'll move this to an Ideas thread, since it belongs there more so than as an Issue/bug.

**Comment by VasilijeBursac** [comment]
Hi everyone!

I wanted to tackle this functionality and contribute my first PR to the Superset repo, but I ran into some design questions and would really appreciate your input.

## Goal
My initial idea was to allow setting a custom loading animation (e.g. `loading.gif`) via a theme token for specifying path to custom `loading.gif` file, similar to how `brandLogoUrl` token works. I thought this would provide a clean and flexible way to swap out the default loader on a per-theme basis.

## What I Tried
I added a new `loadingAnimationUrl: string` token to the `SupersetSpecificTokens` interface in
`/superset-frontend/packages/superset-ui-core/src/theme/types.ts`. Though, maybe `loadingGifUrl` token name would be a better fit in this case.

Then I tried to find all the places where the default `loading.gif` file is used, with the idea of replacing it with `theme.loadingAnimationUrl`. However, I quickly ran into a problem:

The loader is used in both the Jinja base template (`superset/templates/superset/spa.html`) and also in React components (e.g. `ChartLayer.tsx`, `DatasetPanel.tsx`)

## My Dilemma
Since these are two separate rendering layers (Jinja + React), I'm unsure what would be the best approach to share the custom `loading.gif` path between them.

**Here are two approaches I considered:**

**Option 1: Custom Theme Token (`theme.loadingAnimationUrl`):**
- **Pros:** Keeps the config fully within the theme system. React components can easily access it.
- **Cons:** How can we make this value accessible in the base Jinja template (spa.html)?

**Option 2: Additional config in `superset_config.py` (similar to setting custom favicon or custom fonts):**
- **Pros:** Simple to access from Jinja via config.
- **Cons:** How can we access it in the frontend React components? Also, it is less flexible, since it wouldn't support per-theme custom loaders — just one global .gif file per app.

## Setting loader position
As @danielvodnanskyextconradcom stated, it might also be useful to allow custom positioning of the loading gif via the theme tokens. I assume that it could probably be realized according to the same principle as the path itself, but due to the very nature and complexity of setting the position, this is probably a bit more difficult and more cumbersome to do through theme.

Would love to hear your thoughts. Am I on the right track? Maybe @mistercrunch, @rusackas  can chime in with their insights or advice?

Thanks in advance!

**Comment by mistercrunch** [comment]
@rusackas , we have something in-the-works around this I believe, not sure where exactly we're at. Let us know if you're looking for something else to take on!

From what I remember it was animated and theme-aware, svg-based. Looked pretty neat. Can't wait to have that in the app!

**Reply by VasilijeBursac** [reply]
Thank you so much for the quick replies and your engagement! Really appreciate it.

It’s great to hear you’re working on something around this. I’m still learning about the Superset architecture and the way things work in it, but once I get a better understanding and find a place where I can contribute, I’ll definitely do my best to help out!

I’m excited to see this new loading solution in the app as well. Sounds like it’ll be a great addition!

**Reply by mistercrunch** [reply]
I highly recommend giving Claude Code a shot for onboarding, we have a helpful `CLAUDE.md` in this repo that's effectively a system prompt regarding this repo for work with CC, and it does a really good job at discovery on the repo. Prompts like "which part of the codebase does X" or "if I was to implement {feature}, how should I go about it?"

**Comment by jvo05** [comment]
Hello, we are currently searching for a solution to customize the loading spinner. Did something changed in the meanwhile, so it is possible to change the loading animation?

**Reply by VasilijeBursac** [reply]
Loading spinner customization via themes has already been implemented and merged (https://github.com/apache/superset/pull/34764). However, to the best of my knowledge, it is not included in the Superset 6.0 release and is planned for Superset 6.1.

For now, the available workaround is to replace the existing `loading.gif` file with your custom `loading.gif` file (keeping the same filename) in the static assets directory.

**Reply by jvo05** [reply]
Ah okay. Tank you really much for the fast response. Im pretty sure the workaorund should work for us till Superset 6.1

**Reply by VasilijeBursac** [reply]
> Ah okay. Tank you really much for the fast response. Im pretty sure the workaorund should work for us till Superset 6.1

You’re welcome! Yes, I’m already using that approach by replacing the `loading.gif`, and it definitely works as expected.
If you have any further questions regarding this workaround, feel free to ask.
